File: Python3/_20.py
class Solution:
    def isValid(self, s: str) -> bool:
        toClose = []

        for char in s:
            if char in ['(', '{', '[']:
                toClose.append(char)
                continue
                
            if toClose:
                popped = toClose.pop()
                # Matched incorrectly: invalid
                if ((popped == '(' and char != ')') or 
                    (popped == '[' and char != ']') or
                    (popped == '{' and char != '}')):
                    return False

            else: # Close-token left hanging: invalid
                return False

        # Check all found open-tokens are closed
        return not toClose
                

        

File: Cpp/_20.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> toClose;

        for (char c : s) {

            // open- token: push
            // close-token: this switch, confirm there is close-token
            switch(c) {
                case '(':
                case '{':
                case '[':
                    toClose.push(c);
                    continue;

                case ')':
                case '}':
                case ']':
                    if (toClose.empty()) return false;
            };

            //this switch, now confirm existing close-token matches
            switch(c) {
                case ')':
                    if (toClose.top() != '(') return false;
                    break;

                case '}':
                    if (toClose.top() != '{') return false;
                    break;

                case ']':
                    if (toClose.top() != '[') return false;
                    break;
            }

            toClose.pop();
        }

        // Check all found open-tokens are closed
        return toClose.empty();
    }
};